bblfsh/pyuast.cc

```diff
@@ -1,272 +1,760 @@
+  return ctx->lookup(node);
 }
 
-static uint32_t PositionValue(const void* node, const char *prop, const char *field) {
-  PyObject *position = AttributeValue(node, prop);
-  if (!position) {
-    return 0;
-  }
-
-  PyObject *offset = AttributeValue(position, field);
-  Py_DECREF(position);
-  uint32_t retval = 0;
+// PyUastIterExt_dealloc destroys an iterator.
+static void PyUastIterExt_dealloc(PyObject *self) {
+  auto it = (PyUastIterExt *)self;
+  delete(it->iter);
 
-  if (offset) {
-    retval = (uint32_t)PyLong_AsUnsignedLong(offset);
-    Py_DECREF(offset);
+  if (it->freeCtx && it->ctx) {
+    delete(it->ctx);
   }
-  return retval;
+  it->freeCtx = false;
+  it->ctx = nullptr;
+  Py_TYPE(self)->tp_free(self);
 }
 
-/////////////////////////////////////
-/////////// Node Interface //////////
-/////////////////////////////////////
-
-extern "C"
-{
-  static bool HasStartOffset(const void *node) {
-    return HasAttribute(node, "start_position");
-  }
+typedef struct {
+  PyObject_HEAD
+  ContextExt *p;
+} PythonContextExt;
 
-  static uint32_t StartOffset(const void *node) {
-    return PositionValue(node, "start_position", "offset");
-  }
+static void PythonContextExt_dealloc(PyObject *self) {
+  delete(((PythonContextExt *)self)->p);
+  Py_TYPE(self)->tp_free(self);
+}
 
-  static bool HasStartLine(const void *node) {
-    return HasAttribute(node, "start_position");
-  }
+// PythonContextExt_root returns a root node associated with this context.
+// Returns a new reference.
+static PyObject *PythonContextExt_root(PythonContextExt *self, PyObject *Py_UNUSED(ignored)) {
+    return self->p->RootNode();
+}
 
-  static uint32_t StartLine(const void *node) {
-    return PositionValue(node, "start_position", "line");
-  }
+// PythonContextExt_load returns a root node converted to Python object.
+// Returns a new reference.
+static PyObject *PythonContextExt_load(PythonContextExt *self, PyObject *Py_UNUSED(ignored)) {
+    PyObject* root = PythonContextExt_root(self, nullptr);
+    return PyNodeExt_load((PyNodeExt*)root, nullptr);
+}
 
-  static bool HasStartCol(const void *node) {
-    return HasAttribute(node, "start_position");
-  }
+// PythonContextExt_filter filters UAST.
+// Returns a new reference.
+static PyObject *PythonContextExt_filter(PythonContextExt *self, PyObject *args, PyObject *kwargs) {
+    char* kwds[] = {(char*)"query", (char*)"node", NULL};
+    const char *query = nullptr;
+    PyObject *node = nullptr;
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "s|O", kwds, &query, &node))
+      return nullptr;
+
+    PyObject* it = nullptr;
+    try {
+        it = self->p->Filter(node, query);
+    } catch (const std::exception& e) {
+        PyErr_SetString(PyExc_RuntimeError, e.what());
+    }
+    return it;
+}
 
-  static uint32_t StartCol(const void *node) {
-    return PositionValue(node, "start_position", "col");
-  }
+// PythonContextExt_filter serializes UAST.
```

Comment:
But the function is `PythonContextExt_encode` 🤔

Reply:
Fixed in next commit.